ui/src/components/form/multi-select-field.tsx

```diff
@@ -0,0 +1,126 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+import React from 'react';
+import {
+  FieldPathByValue,
+  FieldPathValue,
+  FieldValues,
+  UseFormReturn,
+} from 'react-hook-form';
+
+import { Checkbox } from '@/components/ui/checkbox';
+import {
+  FormControl,
+  FormDescription,
+  FormField,
+  FormItem,
+  FormLabel,
+  FormMessage,
+} from '@/components/ui/form';
+import { Label } from '@/components/ui/label';
+import { Separator } from '@/components/ui/separator';
+
+type MultiSelectFieldProps<
+  TFieldValues extends FieldValues,
+  TName extends FieldPathByValue<TFieldValues, Array<string>>,
+> = {
+  form: UseFormReturn<TFieldValues, TName>;
+  name: TName;
+  label?: string;
+  description?: React.ReactNode;
+  options: readonly { id: string; label: string }[];
+};
+
+export const MultiSelectField = <
+  TFieldValues extends FieldValues,
+  TName extends FieldPathByValue<TFieldValues, Array<string>>,
+>({
+  form,
+  name,
+  label,
+  description,
+  options,
+}: MultiSelectFieldProps<TFieldValues, TName>) => {
+  return (
+    <FormField
+      control={form.control}
+      name={name}
+      render={({ field }) => (
+        <FormItem className='mb-4 flex flex-col justify-between rounded-lg border p-4'>
+          <FormLabel>{label}</FormLabel>
+          <FormDescription className='pb-4'>{description}</FormDescription>
+          <div className='flex items-center space-x-3'>
+            <Checkbox
+              id='check-all-items'
+              checked={
+                options.every((option) =>
+                  form.getValues(name).includes(option.id)
+                )
+                  ? true
+                  : options.some((option) =>
+                        form.getValues(name).includes(option.id)
+                      )
+                    ? 'indeterminate'
+                    : false
+              }
+              onCheckedChange={(checked) => {
+                const enabledItems = checked
+                  ? options.map((option) => option.id)
+                  : [];
+                form.setValue(
+                  name,
+                  // Here, we use the `FieldPathValue` type to ensure that the value is an array of strings.
+                  enabledItems as FieldPathValue<TFieldValues, TName>
```

Comment:
Maybe something like:

TypeScript doesn't get this, but `TName extends FieldPathByValue<TFieldValues, Array<string>>`, so the field behind TName is always an `Array<string>` and `options.map((option) => option.id)` is also `Array<string>`, so this type cast is safe.

Reply:
BTW is there any benefit for using backticks in code comments, or should they just be plain text?

Reply:
To continue, `as` doesn't really ensure anything, but it rather tells TS compiler that whatever it thinks is wrong, here's the correct type. It doesn't do anything to the value, so if you do `const niceNumber = "I'm a number" as number`, you'll get autocompletions and type checking as if it really was a number, but at runtime it's still the same string and can cause bugs.

Reply:
> BTW is there any benefit for using backticks in code comments, or should they just be plain text?

Some editors may display them differently, if you put them in docstrings like:

```typescript
/**
 * This is code: `code`
 */
function foo()
```

For regular inline comments probably no difference, just stylistic choice and communicates that it is code.